packages/flutter/lib/src/material/input_border.dart

```diff
@@ -0,0 +1,346 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+import 'dart:ui' show lerpDouble;
+
+import 'package:flutter/widgets.dart';
+
+/// Defines the appearance of an [InputDecorator]'s border.
+///
+/// An input decorator's border is specified by [InputDecoration.border].
+///
+/// The border is drawn relative to the input decorator's "container" which
+/// is the optionally filled area above the decorator's helper, error,
+/// and counter.
+///
+/// See also:
+///
+///  * [UnderlineInputBorder], the default [InputDecorator] border which
+///    draws a horizontal line at the bottom of the input decorator's container.
+///  * [OutlineInputBorder], an [InputDecorator] border which draws a
+///    rounded rectangle around the input decorator's container.
+///  * [InputDecoration], which is used to configure an [InputDecorator].
+abstract class InputBorder extends ShapeBorder {
+  /// Creates a border for an [InputDecorator].
+  ///
+  /// The [borderSide] parameter must not be null. Applications typically do
+  /// not specify a [borderSide] parameter because the input decorator
+  /// substitutes its own, using [copyWith], based on the current theme and
+  /// [InputDecorator.isFocused].
+  const InputBorder({
+    this.borderSide: BorderSide.none,
+  }) : assert(borderSide != null);
+
+  /// Defines the border line's color and weight.
+  ///
+  /// The [InputDecorator] creates copies of its input border, using [copyWith],
+  /// based on the current theme and [InputDecorator.isFocused].
+  final BorderSide borderSide;
+
+  /// Creates a copy of this input border with the specified `borderSide`.
+  InputBorder copyWith({ BorderSide borderSide });
+}
```

Comment:
I think we should add the following to this class, to avoid needing to hard-code any knowledge about InputBorder's subclasses in any of the other classes, thus allowing people to implement custom InputBorders that are crazier than currently allowed, while also simplifying the rest of the code:

- add an abstract paint() method that has the arguments that OutlineInputBorder has.
- add a property that returns an enum that describes how the padding/baseline should be applied (also covers the isOutlineBorder logic in the paint method).

We should also document in the class docs for InputBorder all the ways that this class differs from ShapeBorder, in particular:

- That it always has a single side style, `borderSide`.
- That copies of the class can be made, applying a different style, using `copyWith`.
- That the `paint` method has additional arguments.
- What the enum property described above is about.

Reply:
InputBorder now has an abstract `isOutline` getter that's used by the renderer to configure the layout. The renderer tests this boolean instead of testing the border's type.

I updated the abstract methods and docs as you suggested.